overlord/servicestate/quota_control.go

```diff
@@ -42,9 +52,65 @@ func checkSystemdVersion() {
 
 func init() {
 	checkSystemdVersion()
+	checkMemoryCGroupEnabled()
+}
+
+func checkMemoryCGroupEnabled() {
+	memoryCGroupError = memoryCGroupEnabled()
+}
+
+// added to enable path update for testing purposes
+func setCGroupsFilePath(path string) {
+	cgroupsFilePath = path
+	checkMemoryCGroupEnabled()
+}
+
+//	since the control groups can be enabled/disabled without the kernel config the only
+//	way to identify the status of memory control groups is via /proc/cgroups
+//	"cat /proc/cgroups | grep memory" returns the active status of memory control group
+//	and the 3rd parameter is the status
+//	0 => false => disabled
+//	1 => true => enabled
+func memoryCGroupEnabled() error {
+	cgroupsFile, err := os.Open(cgroupsFilePath)
+	if err != nil {
+		return fmt.Errorf("cannot open cgroups file: %v", err)
+	}
+	defer cgroupsFile.Close()
+	scanner := bufio.NewScanner(cgroupsFile)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "memory\t") {
+			memoryCgroupValues := strings.Fields(line)
+			if len(memoryCgroupValues) >= 4 { // assuming any increase in size will lead to values added to the end
+				isMemoryEnabled, err := strconv.ParseBool(memoryCgroupValues[3])
+				if err != nil {
+					return fmt.Errorf("cannot parse memory control group status: %v", err)
+				}
+				if !isMemoryEnabled {
+					return fmt.Errorf("cannot retrieve quota information, memory cgroup is disabled on this system")
+				}
+				return nil
+			}
+			// change in size, should investigate the new structure
+			return fmt.Errorf("cannot parse memory control group configuration")
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("cannot read %v contents: %v", cgroupsFilePath, err)
+	}
+
+	// no errors so far but the only path here is the cgroups file without the memory line
+	return fmt.Errorf("cannot retrieve memory cgroup configuration, it is not available in the kernel config")
```

Comment:
no test case hits this code

Reply:
added the test case for missing memory line in configuration